pypreprocess/nipype_preproc_fsl_utils.py

```diff
@@ -180,3 +181,174 @@ def do_subject_preproc(subject_id,
+
+    Parameters
+    ----------
+    ap_realig_img: string
+        path to img using negative phase-encode blips. Anterior --> Posterior
+        the image should be already realigned
+
+    pa_realig_img: string
+        path to img using positive phase-encode blips. Posterior --> Anterior
+        the image should be already realigned
+
+    func_imgs: list of string
+        functional imgs path, that will be corrected with topup
+
+    total_readout_times: tuple of floats, optional (default None)
+        the first float corresponds to the total_readout_times of the ap img
+        and the second float to the pa img. Its crucial to provide these if
+        they are different. The times should be provided in seconds.
+
+    memory: Nipype `Memory` object, optional (default None)
+        if given, then caching will be enabled
+
+    tmp_dir: string, optional (default None)
+        temporary directory to store acquisition parameters configuration file
+        for topup. Must be provided in case memory is provided.
+    """
+    if total_readout_times is None:
+        total_readout_times = (1., 1.)  # Check FSL documentation explanation
+
+    if memory is not None:
+        if tmp_dir is None:
+            raise('Temporary directory was not provided with memory object')
+        # Merge AP and PA images
+        merge = memory.cache(fsl.Merge)
+        appa_merged = merge(in_files=[ap_realig_img, pa_realig_img],
+                            dimension='t')
+
+        # Compute topup transformation
+        acq_param_file = os.path.join(tmp_dir, 'acq_param.txt')
+        if not os.path.isfile(acq_param_file):
+            with open(acq_param_file, 'w') as acq:
+                content = '0 -1 0 {0:6.5f} \n0 1 0 {0:6.5f}'
+                acq.write(content % total_readout_times)
+
+        topup = memory.cache(fsl.TOPUP)
+        correction = topup(in_file=appa_merged.outputs.merged_file,
+                           encoding_file=acq_param_file,
+                           output_type='NIFTI', out_base='APPA_DefMap',
+                           out_field='sanitycheck_DefMap',
+                           out_corrected='sanitycheck_unwarped_B0')
+
+        # Apply topup correction to images
+        fieldcoef = correction.outputs.out_fieldcoef
+        movpar = correction.outputs.out_movpar
+        applytopup = memory.cache(fsl.ApplyTOPUP)
+        return [applytopup(in_files=img, encoding_file=acq_param_file,
+                           in_index=[2], in_topup_fieldcoef=fieldcoef,
+                           in_topup_movpar=movpar, output_type='NIFTI',
+                           method='jac') for img in func_imgs]
+
+    else:
+        # Merge AP and PA images
+        appa_merged = fsl.Merge(in_files=[ap_realig_img, pa_realig_img],
+                                dimension='t').run()
+
+        # Compute topup transformation
+        acq_param_file = os.path.join('/tmp', 'pypreprocess_topup',
```

Comment:
On Fri, May 27, 2016 at 6:04 PM, mrahim <notifications@github.com> wrote:

> In pypreprocess/nipype_preproc_fsl_utils.py
> <https://github.com/neurospin/pypreprocess/pull/229#discussion_r64928000>:
>
> > +        # Apply topup correction to images
> > +        fieldcoef = correction.outputs.out_fieldcoef
> > +        movpar = correction.outputs.out_movpar
> > +        applytopup = memory.cache(fsl.ApplyTOPUP)
> > +        return [applytopup(in_files=img, encoding_file=acq_param_file,
> > +                           in_index=[2], in_topup_fieldcoef=fieldcoef,
> > +                           in_topup_movpar=movpar, output_type='NIFTI',
> > +                           method='jac') for img in func_imgs]
> > +
> > +    else:
> > +        # Merge AP and PA images
> > +        appa_merged = fsl.Merge(in_files=[ap_realig_img, pa_realig_img],
> > +                                dimension='t').run()
> > +
> > +        # Compute topup transformation
> > +        acq_param_file = os.path.join('/tmp', 'pypreprocess_topup',
>
> is it okay to create this directory ?
>
No. We should store the parameters in the subject's output directory
(subject_data.output_dir).

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/neurospin/pypreprocess/pull/229/files/d9eba52dc19c47e59073e19f282f4d83ce9d3631#r64928000>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAms1BP-syhHiAUENWCPtrxJxOImL1oiks5qFxYRgaJpZM4IooTp>
> .
>



-- 
Elvis Dohmatob,
*sent from my  potato...*
http://dohmatob.blogspot.fr/, https://team.inria.fr/parietal/elvis/,
https://github.com/dohmatob